<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/xsl" href="../assets/xml/rss.xsl" media="all"?><rss version="2.0" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:atom="http://www.w3.org/2005/Atom"><channel><title>The Cloistered Monkey (Posts about upgrading)</title><link>https://necromuralist.github.io/</link><description></description><atom:link href="https://necromuralist.github.io/categories/upgrading.xml" rel="self" type="application/rss+xml"></atom:link><language>en</language><copyright>Contents © 2025 &lt;a href="mailto:cloisteredmonkey.jmark@slmail.me"&gt;Cloistered Monkey&lt;/a&gt; 
&lt;div id="license"xmlns:cc="http://creativecommons.org/ns#" &gt;This work is licensed under
&lt;a href="http://creativecommons.org/licenses/by/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display:inline-block;"&gt;CC BY 4.0
&lt;img src="https://mirrors.creativecommons.org/presskit/icons/cc.svg?ref=chooser-v1"&gt;
&lt;img src="https://mirrors.creativecommons.org/presskit/icons/by.svg?ref=chooser-v1"&gt;&lt;/a&gt;
&lt;/div&gt;
</copyright><lastBuildDate>Sat, 19 Apr 2025 22:55:47 GMT</lastBuildDate><generator>Nikola (getnikola.com)</generator><docs>http://blogs.law.harvard.edu/tech/rss</docs><item><title>Noble Numbat and the Broken Virtual Environments</title><link>https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/</link><dc:creator>Cloistered Monkey</dc:creator><description>&lt;div id="table-of-contents" role="doc-toc"&gt;
&lt;h2&gt;Table of Contents&lt;/h2&gt;
&lt;div id="text-table-of-contents" role="doc-toc"&gt;
&lt;ul&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#orge191ef7"&gt;What is this about?&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#org0c7176f"&gt;What Happened&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#orgad64f4b"&gt;Short Version&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#orga5f4cd8"&gt;The First Hint (pipx)&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#orgafa04e1"&gt;The Broken Virtual Environments&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#org62c9b7c"&gt;Then a Nikola Error&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#org934eba2"&gt;An Ape-Iron Problem (no distutils)&lt;/a&gt;&lt;/li&gt;
&lt;li&gt;&lt;a href="https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/#orgfeb54ce"&gt;Links&lt;/a&gt;&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-orge191ef7" class="outline-2"&gt;
&lt;h2 id="orge191ef7"&gt;What is this about?&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orge191ef7"&gt;
&lt;p&gt;
These are notes I made when I was fixing some errors that occurred after updating Ubuntu to 24.04. It's been over a year so I don't really remember writing this, but, here it is.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org0c7176f" class="outline-2"&gt;
&lt;h2 id="org0c7176f"&gt;What Happened&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org0c7176f"&gt;
&lt;ul class="org-ul"&gt;
&lt;li&gt;I updated Ubuntu&lt;/li&gt;
&lt;li&gt;Something raised an obscure error when I changed into one of my git repositories&lt;/li&gt;
&lt;li&gt;It turned out that the upgrade broke the python virtual environment&lt;/li&gt;
&lt;li&gt;Re-setting-up using the newer version of python that came with the upgrade fixed things.&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-orgad64f4b" class="outline-2"&gt;
&lt;h2 id="orgad64f4b"&gt;Short Version&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orgad64f4b"&gt;
&lt;ul class="org-ul"&gt;
&lt;li&gt;Upgrading to Ubuntu 24.04 made python changes that broke all my virtual enviroments&lt;/li&gt;
&lt;li&gt;Reinstalled &lt;code&gt;pipx&lt;/code&gt; using apt&lt;/li&gt;
&lt;li&gt;&lt;code&gt;vf ls --details&lt;/code&gt; shows the python versions for all (virtualfish created) virtual environments (and which are broken)&lt;/li&gt;
&lt;li&gt;&lt;code&gt;vf upgrade --rebuild&lt;/code&gt; will upgrade the currently active virtual environment (may take more than one run to get past errors)&lt;/li&gt;
&lt;li&gt;&lt;code&gt;pipx reinstall-all&lt;/code&gt; fixed the missing &lt;code&gt;nikola&lt;/code&gt; and &lt;code&gt;pygmentize&lt;/code&gt; errors (and probably others)&lt;/li&gt;
&lt;li&gt;&lt;code&gt;pip install setuptools --upgrade&lt;/code&gt; in the environment re-setup &lt;code&gt;distutils&lt;/code&gt; which was removed from pthyon 3.12 (which Ubuntu Installed)&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-orga5f4cd8" class="outline-2"&gt;
&lt;h2 id="orga5f4cd8"&gt;The First Hint (pipx)&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orga5f4cd8"&gt;
&lt;p&gt;
I don't remember exactly what cued me in to what was going wrong, but somewhere in trying to figure out what was wrong I found that running &lt;a href="https://github.com/frostming/legit"&gt;legit&lt;/a&gt; after updating to Ubuntu 24.04 returned an error saying that &lt;code&gt;legit&lt;/code&gt; wasn't found. It took a while to figure out where &lt;code&gt;legit&lt;/code&gt; was supposed to be coming from, but it turned out I had installed it with &lt;a href="https://pipx.pypa.io/latest/installation/"&gt;pipx&lt;/a&gt; and had to re-install it to get it working.
&lt;/p&gt;

&lt;p&gt;
I don't remember exactly what I did anymore, but based on the command history it looks like:
&lt;/p&gt;

&lt;ol class="org-ol"&gt;
&lt;li&gt;Re-install &lt;code&gt;pipx&lt;/code&gt; (I purged the original installation that had been installed using &lt;code&gt;pip&lt;/code&gt; and instead used &lt;code&gt;apt&lt;/code&gt;)&lt;/li&gt;
&lt;li&gt;Re-install &lt;code&gt;legit&lt;/code&gt; using the updated &lt;code&gt;pipx&lt;/code&gt;&lt;/li&gt;
&lt;/ol&gt;

&lt;p&gt;
So, was &lt;code&gt;pipx&lt;/code&gt; the problem? Well, partly.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-orgafa04e1" class="outline-2"&gt;
&lt;h2 id="orgafa04e1"&gt;The Broken Virtual Environments&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orgafa04e1"&gt;
&lt;p&gt;
The next problem came when I tried to build the HTML for this site.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;nikola build
&lt;/pre&gt;&lt;/div&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="n"&gt;Traceback&lt;/span&gt; &lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="n"&gt;most&lt;/span&gt; &lt;span class="n"&gt;recent&lt;/span&gt; &lt;span class="n"&gt;call&lt;/span&gt; &lt;span class="n"&gt;last&lt;/span&gt;&lt;span class="p"&gt;):&lt;/span&gt;
  &lt;span class="n"&gt;File&lt;/span&gt; &lt;span class="s2"&gt;"/home/hades/.virtualenvs/necromuralist.github.io/bin/nikola"&lt;/span&gt;&lt;span class="p"&gt;,&lt;/span&gt; &lt;span class="n"&gt;line&lt;/span&gt; &lt;span class="mi"&gt;5&lt;/span&gt;&lt;span class="p"&gt;,&lt;/span&gt; &lt;span class="ow"&gt;in&lt;/span&gt; &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;&lt;span class="n"&gt;module&lt;/span&gt;&lt;span class="o"&gt;&amp;gt;&lt;/span&gt;
    &lt;span class="kn"&gt;from&lt;/span&gt; &lt;span class="nn"&gt;nikola.__main__&lt;/span&gt; &lt;span class="kn"&gt;import&lt;/span&gt; &lt;span class="n"&gt;main&lt;/span&gt;
&lt;span class="ne"&gt;ModuleNotFoundError&lt;/span&gt;&lt;span class="p"&gt;:&lt;/span&gt; &lt;span class="n"&gt;No&lt;/span&gt; &lt;span class="n"&gt;module&lt;/span&gt; &lt;span class="n"&gt;named&lt;/span&gt; &lt;span class="s1"&gt;'nikola'&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
For some reason the &lt;code&gt;nikola&lt;/code&gt; executable was missing. Since I was in a virtual environment I tried to re-install nikola but using &lt;code&gt;pip&lt;/code&gt; gave another error.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;pip --version
&lt;/pre&gt;&lt;/div&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="n"&gt;Traceback&lt;/span&gt; &lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="n"&gt;most&lt;/span&gt; &lt;span class="n"&gt;recent&lt;/span&gt; &lt;span class="n"&gt;call&lt;/span&gt; &lt;span class="n"&gt;last&lt;/span&gt;&lt;span class="p"&gt;):&lt;/span&gt;
  &lt;span class="n"&gt;File&lt;/span&gt; &lt;span class="s2"&gt;"/home/hades/.virtualenvs/necromuralist.github.io/bin/pip"&lt;/span&gt;&lt;span class="p"&gt;,&lt;/span&gt; &lt;span class="n"&gt;line&lt;/span&gt; &lt;span class="mi"&gt;5&lt;/span&gt;&lt;span class="p"&gt;,&lt;/span&gt; &lt;span class="ow"&gt;in&lt;/span&gt; &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;&lt;span class="n"&gt;module&lt;/span&gt;&lt;span class="o"&gt;&amp;gt;&lt;/span&gt;
    &lt;span class="kn"&gt;from&lt;/span&gt; &lt;span class="nn"&gt;pip._internal.cli.main&lt;/span&gt; &lt;span class="kn"&gt;import&lt;/span&gt; &lt;span class="n"&gt;main&lt;/span&gt;
&lt;span class="ne"&gt;ModuleNotFoundError&lt;/span&gt;&lt;span class="p"&gt;:&lt;/span&gt; &lt;span class="n"&gt;No&lt;/span&gt; &lt;span class="n"&gt;module&lt;/span&gt; &lt;span class="n"&gt;named&lt;/span&gt; &lt;span class="s1"&gt;'pip'&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
So, not only was &lt;code&gt;nikola&lt;/code&gt; missing but so was &lt;code&gt;pip&lt;/code&gt;, which I thought was a part of the python installation… I looked at what was in the virtualenvironment -
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;ls ~/.virtualenvs/necromuralist.github.io/lib
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
And saw that it had this python version:
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;python3.11/
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
But when I checked the version of the python I was using it gave this.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;python --version
&lt;/pre&gt;&lt;/div&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;Python 3.12.3
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
So it looked like the ubuntu update created a bit of a mismatch. It turns out that &lt;a href="https://virtualfish.readthedocs.io/en/latest/"&gt;virtualfish&lt;/a&gt; has a command to list what virtual environments you have set up (&lt;code&gt;vf ls&lt;/code&gt;) and if you pass it the &lt;code&gt;--details=&lt;/code&gt; option it will show you the python version the environment is using.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;vf ls --details
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
It listed all my environments but in particular it showed this for the environment for this site's repository.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;necromuralist.github.io &lt;span class="o"&gt;(&lt;/span&gt;broken&lt;span class="o"&gt;)&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
&lt;code&gt;virtualfish&lt;/code&gt; also has a command to upgrade the currently active environment so I tried running it.
&lt;/p&gt;

&lt;p&gt;
Running this with the &lt;i&gt;necromuralist&lt;/i&gt; environment active:
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;vf upgrade --rebuild
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Gave this error.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;Collecting &lt;span class="nv"&gt;PyYAML&lt;/span&gt;&lt;span class="o"&gt;==&lt;/span&gt;6.0
  Downloading PyYAML-6.0.tar.gz &lt;span class="o"&gt;(&lt;/span&gt;&lt;span class="m"&gt;124&lt;/span&gt; kB&lt;span class="o"&gt;)&lt;/span&gt;
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.0/125.0 kB 3.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did &lt;span class="k"&gt;not&lt;/span&gt; run successfully.
  │ &lt;span class="k"&gt;exit&lt;/span&gt; code: 1
  ╰─&amp;gt; &lt;span class="o"&gt;[&lt;/span&gt;&lt;span class="m"&gt;54&lt;/span&gt; lines of output&lt;span class="o"&gt;]&lt;/span&gt;
      running egg_info
      writing lib/PyYAML.egg-info/PKG-INFO
      writing dependency_links to lib/PyYAML.egg-info/dependency_links.txt
      writing top-level names to lib/PyYAML.egg-info/top_level.txt
      Traceback &lt;span class="o"&gt;(&lt;/span&gt;most recent call last&lt;span class="o"&gt;)&lt;/span&gt;:
        File &lt;span class="s2"&gt;"/home/hades/.virtualenvs/necromuralist.github.io/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py"&lt;/span&gt;, line 353, &lt;span class="k"&gt;in&lt;/span&gt; &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;module&amp;gt;
          main&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/home/hades/.virtualenvs/necromuralist.github.io/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py"&lt;/span&gt;, line 335, &lt;span class="k"&gt;in&lt;/span&gt; main
          json_out&lt;span class="o"&gt;[&lt;/span&gt;&lt;span class="s1"&gt;'return_val'&lt;/span&gt;&lt;span class="o"&gt;]&lt;/span&gt; &lt;span class="o"&gt;=&lt;/span&gt; hook&lt;span class="o"&gt;(&lt;/span&gt;**hook_input&lt;span class="o"&gt;[&lt;/span&gt;&lt;span class="s1"&gt;'kwargs'&lt;/span&gt;&lt;span class="o"&gt;])&lt;/span&gt;
                                   &lt;span class="o"&gt;^^^^^^^^^^^^^^^^^^^^^^^^^^^^&lt;/span&gt;
        File &lt;span class="s2"&gt;"/home/hades/.virtualenvs/necromuralist.github.io/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py"&lt;/span&gt;, line 118, &lt;span class="k"&gt;in&lt;/span&gt; get_requires_for_build_wheel
          &lt;span class="k"&gt;return&lt;/span&gt; hook&lt;span class="o"&gt;(&lt;/span&gt;config_settings&lt;span class="o"&gt;)&lt;/span&gt;
                 &lt;span class="o"&gt;^^^^^^^^^^^^^^^^^^^^^&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/build_meta.py"&lt;/span&gt;, line 325, &lt;span class="k"&gt;in&lt;/span&gt; get_requires_for_build_wheel
          &lt;span class="k"&gt;return&lt;/span&gt; self._get_build_requires&lt;span class="o"&gt;(&lt;/span&gt;config_settings, &lt;span class="nv"&gt;requirements&lt;/span&gt;&lt;span class="o"&gt;=[&lt;/span&gt;&lt;span class="s1"&gt;'wheel'&lt;/span&gt;&lt;span class="o"&gt;])&lt;/span&gt;
                 &lt;span class="o"&gt;^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/build_meta.py"&lt;/span&gt;, line 295, &lt;span class="k"&gt;in&lt;/span&gt; _get_build_requires
          self.run_setup&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/build_meta.py"&lt;/span&gt;, line 311, &lt;span class="k"&gt;in&lt;/span&gt; run_setup
          &lt;span class="nb"&gt;exec&lt;/span&gt;&lt;span class="o"&gt;(&lt;/span&gt;code, locals&lt;span class="o"&gt;())&lt;/span&gt;
        File &lt;span class="s2"&gt;"&amp;lt;string&amp;gt;"&lt;/span&gt;, line 288, &lt;span class="k"&gt;in&lt;/span&gt; &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;module&amp;gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/__init__.py"&lt;/span&gt;, line 103, &lt;span class="k"&gt;in&lt;/span&gt; setup
          &lt;span class="k"&gt;return&lt;/span&gt; distutils.core.setup&lt;span class="o"&gt;(&lt;/span&gt;**attrs&lt;span class="o"&gt;)&lt;/span&gt;
                 &lt;span class="o"&gt;^^^^^^^^^^^^^^^^^^^^^^^^^^^^^&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/core.py"&lt;/span&gt;, line 184, &lt;span class="k"&gt;in&lt;/span&gt; setup
          &lt;span class="k"&gt;return&lt;/span&gt; run_commands&lt;span class="o"&gt;(&lt;/span&gt;dist&lt;span class="o"&gt;)&lt;/span&gt;
                 &lt;span class="o"&gt;^^^^^^^^^^^^^^^^^^&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/core.py"&lt;/span&gt;, line 200, &lt;span class="k"&gt;in&lt;/span&gt; run_commands
          dist.run_commands&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/dist.py"&lt;/span&gt;, line 969, &lt;span class="k"&gt;in&lt;/span&gt; run_commands
          self.run_command&lt;span class="o"&gt;(&lt;/span&gt;cmd&lt;span class="o"&gt;)&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/dist.py"&lt;/span&gt;, line 968, &lt;span class="k"&gt;in&lt;/span&gt; run_command
          super&lt;span class="o"&gt;()&lt;/span&gt;.run_command&lt;span class="o"&gt;(&lt;/span&gt;&lt;span class="nb"&gt;command&lt;/span&gt;&lt;span class="o"&gt;)&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/dist.py"&lt;/span&gt;, line 988, &lt;span class="k"&gt;in&lt;/span&gt; run_command
          cmd_obj.run&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py"&lt;/span&gt;, line 321, &lt;span class="k"&gt;in&lt;/span&gt; run
          self.find_sources&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py"&lt;/span&gt;, line 329, &lt;span class="k"&gt;in&lt;/span&gt; find_sources
          mm.run&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py"&lt;/span&gt;, line 550, &lt;span class="k"&gt;in&lt;/span&gt; run
          self.add_defaults&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/command/egg_info.py"&lt;/span&gt;, line 588, &lt;span class="k"&gt;in&lt;/span&gt; add_defaults
          sdist.add_defaults&lt;span class="o"&gt;(&lt;/span&gt;self&lt;span class="o"&gt;)&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/command/sdist.py"&lt;/span&gt;, line 102, &lt;span class="k"&gt;in&lt;/span&gt; add_defaults
          super&lt;span class="o"&gt;()&lt;/span&gt;.add_defaults&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/command/sdist.py"&lt;/span&gt;, line 250, &lt;span class="k"&gt;in&lt;/span&gt; add_defaults
          self._add_defaults_ext&lt;span class="o"&gt;()&lt;/span&gt;
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/command/sdist.py"&lt;/span&gt;, line 335, &lt;span class="k"&gt;in&lt;/span&gt; _add_defaults_ext
          self.filelist.extend&lt;span class="o"&gt;(&lt;/span&gt;build_ext.get_source_files&lt;span class="o"&gt;())&lt;/span&gt;
                               &lt;span class="o"&gt;^^^^^^^^^^^^^^^^^^^^^^^^^^^^&lt;/span&gt;
        File &lt;span class="s2"&gt;"&amp;lt;string&amp;gt;"&lt;/span&gt;, line 204, &lt;span class="k"&gt;in&lt;/span&gt; get_source_files
        File &lt;span class="s2"&gt;"/tmp/pip-build-env-1ai8jd66/overlay/lib/python3.12/site-packages/setuptools/_distutils/cmd.py"&lt;/span&gt;, line 107, &lt;span class="k"&gt;in&lt;/span&gt; __getattr__
          raise AttributeError&lt;span class="o"&gt;(&lt;/span&gt;attr&lt;span class="o"&gt;)&lt;/span&gt;
      AttributeError: cython_sources
      &lt;span class="o"&gt;[&lt;/span&gt;&lt;span class="k"&gt;end&lt;/span&gt; of output&lt;span class="o"&gt;]&lt;/span&gt;

  note: This error originates from a subprocess, &lt;span class="k"&gt;and&lt;/span&gt; is likely &lt;span class="k"&gt;not&lt;/span&gt; a problem with pip.
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Which looked like I was in trouble, but it turned out that re-running the upgrade fixed it. One of those magic fixes that would be scary if this was a critical site, I suppose, but which is instead, just another bit of python mystery.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org62c9b7c" class="outline-2"&gt;
&lt;h2 id="org62c9b7c"&gt;Then a Nikola Error&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org62c9b7c"&gt;
&lt;p&gt;
So, with the virtual environment updated to the current python, I waded back into building the site.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;nikola build
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
First this produced a &lt;code&gt;nikola&lt;/code&gt; not found error (updating the virtual environment apparently doesn't fix everything). But since the environment itself was fixed pip was now working so I installed it again, but then building the site produced a different error.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;DependencyError - taskid:render_posts:cache/posts/2ac-updating-a-nikola-shortcode-plugin.html  
ERROR: Task &lt;span class="s1"&gt;'render_posts:cache/posts/2ac-updating-a-nikola-shortcode-plugin.html'&lt;/span&gt; saving success: Dependent file &lt;span class="s1"&gt;'/home/hades/.virtualenvs/necromuralist.github.io/lib/python3.11/site-packa&lt;/span&gt;
&lt;span class="s1"&gt;ges/nikola/data/shortcodes/mako/raw.tmpl'&lt;/span&gt; does &lt;span class="k"&gt;not&lt;/span&gt; exist.
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Re-reinstalling nikola and re-running the build made the error go away, but in the output I noticed some disturbing messages that looked like these.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;.  render_pages:output/posts/destroying-tags-with-beautiful-soup/index.html
line &lt;span class="m"&gt;135&lt;/span&gt; column &lt;span class="m"&gt;56&lt;/span&gt; - Error: &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;module&amp;gt; is &lt;span class="k"&gt;not&lt;/span&gt; recognized!
line &lt;span class="m"&gt;144&lt;/span&gt; column &lt;span class="m"&gt;56&lt;/span&gt; - Error: &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;module&amp;gt; is &lt;span class="k"&gt;not&lt;/span&gt; recognized!
line &lt;span class="m"&gt;154&lt;/span&gt; column &lt;span class="m"&gt;56&lt;/span&gt; - Error: &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;module&amp;gt; is &lt;span class="k"&gt;not&lt;/span&gt; recognized!
This document has errors that must be fixed before
using HTML Tidy to generate a tidied up version.
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
There were are a lot of these errors (which didn't cause nikola to quit so I wouldn't have seen them if I wasn't watching the build messages).
&lt;/p&gt;

&lt;p&gt;
This turned out to be because nikola was encountering errors, trapping them and not failing and then embedding the python error messages into the HTML (and the error messages were causing HTML tidy to fail and then raise a different error message).
&lt;/p&gt;

&lt;p&gt;
This is an example of an error message embedded in the HTML nikola created:
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;&lt;span class="n"&gt;Traceback&lt;/span&gt; &lt;span class="p"&gt;(&lt;/span&gt;&lt;span class="n"&gt;most&lt;/span&gt; &lt;span class="n"&gt;recent&lt;/span&gt; &lt;span class="n"&gt;call&lt;/span&gt; &lt;span class="n"&gt;last&lt;/span&gt;&lt;span class="p"&gt;):&lt;/span&gt;
  &lt;span class="n"&gt;File&lt;/span&gt; &lt;span class="s2"&gt;"/home/hades/.local/bin/pygmentize"&lt;/span&gt;&lt;span class="p"&gt;,&lt;/span&gt; &lt;span class="n"&gt;line&lt;/span&gt; &lt;span class="mi"&gt;5&lt;/span&gt;&lt;span class="p"&gt;,&lt;/span&gt; &lt;span class="ow"&gt;in&lt;/span&gt; &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;&lt;span class="n"&gt;module&lt;/span&gt;&lt;span class="o"&gt;&amp;gt;&lt;/span&gt;
    &lt;span class="kn"&gt;from&lt;/span&gt; &lt;span class="nn"&gt;pygments.cmdline&lt;/span&gt; &lt;span class="kn"&gt;import&lt;/span&gt; &lt;span class="n"&gt;main&lt;/span&gt;
&lt;span class="ne"&gt;ModuleNotFoundError&lt;/span&gt;&lt;span class="p"&gt;:&lt;/span&gt; &lt;span class="n"&gt;No&lt;/span&gt; &lt;span class="n"&gt;module&lt;/span&gt; &lt;span class="n"&gt;named&lt;/span&gt; &lt;span class="s1"&gt;'pygments'&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
The cause for this one was a little more mysterious, but it turned out that I had installed &lt;code&gt;pygmentize&lt;/code&gt; using &lt;code&gt;pipx&lt;/code&gt; so it needed to be updated separately. I decided to use &lt;code&gt;pipx reinstall-all&lt;/code&gt; to get everything updated.
&lt;/p&gt;

&lt;p&gt;
Unfortunately, there were a lot of posts with the &lt;code&gt;pygmentize&lt;/code&gt; errors, and since they aren't registered as errors by nikola, just building the site didn't fix them so I wiped it first and then re-built it.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;nikola clean
nikola build
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
And since there weren't anymore errors I decided to cross my fingers and move on.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-org934eba2" class="outline-2"&gt;
&lt;h2 id="org934eba2"&gt;An Ape-Iron Problem (no distutils)&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-org934eba2"&gt;
&lt;p&gt;
The &lt;code&gt;vf ls --details&lt;/code&gt; command I used to check on the virtual environment for this site also showed that all my other virtual environments were broken so I decided to try and fix the &lt;code&gt;Ape-Iron&lt;/code&gt; repository while it was fresh on my mind. First I rebuilt the environment.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;vf upgrade --rebuild Ape-Iron
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
And of course, it gave me errors.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;Collecting &lt;span class="nv"&gt;numpy&lt;/span&gt;&lt;span class="o"&gt;==&lt;/span&gt;1.24.3
  Downloading numpy-1.24.3.tar.gz &lt;span class="o"&gt;(&lt;/span&gt;10.9 MB&lt;span class="o"&gt;)&lt;/span&gt;
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 7.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
ERROR: Exception:
Traceback &lt;span class="o"&gt;(&lt;/span&gt;most recent call last&lt;span class="o"&gt;)&lt;/span&gt;:
  File &lt;span class="s2"&gt;"/home/hades/.virtualenvs/Ape-Iron/lib/python3.12/site-packages/pip/_internal/cli/base_command.py"&lt;/span&gt;, line 180, &lt;span class="k"&gt;in&lt;/span&gt; exc_logging_wrapper
&lt;/pre&gt;&lt;/div&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;File &lt;span class="s2"&gt;"/tmp/pip-build-env-f7h7lvcc/overlay/lib/python3.12/site-packages/setuptools/__init__.py"&lt;/span&gt;, line 10, &lt;span class="k"&gt;in&lt;/span&gt; &lt;span class="o"&gt;&amp;lt;&lt;/span&gt;module&amp;gt;
    import distutils.core
ModuleNotFoundError: No module named &lt;span class="s1"&gt;'distutils'&lt;/span&gt;
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
According to &lt;a href="https://stackoverflow.com/a/77233866"&gt;this answer&lt;/a&gt; to a question on &lt;i&gt;StackOverflow&lt;/i&gt; &lt;code&gt;distutils&lt;/code&gt; was removed from python 3.12, which is the default with &lt;i&gt;noble-numbat&lt;/i&gt;. The answer suggests installing &lt;a href="https://github.com/pypa/setuptools"&gt;&lt;code&gt;setuptools&lt;/code&gt;&lt;/a&gt;.
&lt;/p&gt;

&lt;p&gt;
So, I tride that.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;pip install setuptools --upgrade
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
Amazingly, the upgrade of Ape-Iron seemed to work.
&lt;/p&gt;

&lt;div class="highlight"&gt;&lt;pre&gt;&lt;span&gt;&lt;/span&gt;vf upgrade --rebuild Ape-Iron
&lt;/pre&gt;&lt;/div&gt;

&lt;p&gt;
At this point I think I just called it a day.
&lt;/p&gt;
&lt;/div&gt;
&lt;/div&gt;

&lt;div id="outline-container-orgfeb54ce" class="outline-2"&gt;
&lt;h2 id="orgfeb54ce"&gt;Links&lt;/h2&gt;
&lt;div class="outline-text-2" id="text-orgfeb54ce"&gt;
&lt;ul class="org-ul"&gt;
&lt;li&gt;&lt;a href="https://virtualfish.readthedocs.io/en/latest/usage.html#upgrading-virtual-environments"&gt;Upgrading Virtual Environments&lt;/a&gt;: Virtualfish Documentation&lt;/li&gt;
&lt;li&gt;&lt;a href="https://github.com/yaml/pyyaml/issues/601"&gt;AttributeError: cython_sources&lt;/a&gt;: A PyYAML bug report that looks related to the error.&lt;/li&gt;
&lt;/ul&gt;
&lt;/div&gt;
&lt;/div&gt;</description><category>pipx</category><category>ubuntu</category><category>upgrading</category><category>virtualfish</category><guid>https://necromuralist.github.io/posts/noble-numbat-and-the-broken-virtual-environments/</guid><pubDate>Mon, 27 May 2024 23:21:26 GMT</pubDate></item></channel></rss>